cq/cqlearners.py

```diff
@@ -0,0 +1,196 @@
+        isNecessary, joint_states_including_local_state = self.__isCoordinationNecessary(local_state)
+
+        if (isNecessary and (global_state in joint_states_including_local_state) 
+          #  and self.coordination_joint_states_confidence[global_state] > np.median(list(self.coordination_joint_states_confidence.values()))
+          ):
+            action = self.global_action_selector.selectAction(global_state)
+            is_joint = True
+            self.__incrementCoordinationJointStateConfidence(joint_states_including_local_state)
+        else:
+            action = self.local_action_selector.selectAction(local_state)
+            self.__decrementCoordinationJointStateConfidence(joint_states_including_local_state)
+            is_joint = False
+            
+        return (action, is_joint) 
+    
+    def __isCoordinationNecessary(self, state):
+        joint_actions = [joint_state for joint_state in self.coordination_joint_states if joint_state[0] == state]
+        return len(joint_actions) > 0, joint_actions
+            
+    def __incrementCoordinationJointStateConfidence(self, joint_states):
+        for js in joint_states:            
+            self.coordination_joint_states_confidence[js] += 1
+        
+    def __decrementCoordinationJointStateConfidence(self, joint_states):
+        for js in joint_states:            
+            self.coordination_joint_states_confidence[js] -= 13
+            # larger discount valuese preferred for game1 (ex: 13) and larger for game2 (30)
+            
+            if (self.coordination_joint_states_confidence[js] < 5):
+                self.coordination_joint_states.remove(js)
+                del self.coordination_joint_states_confidence[js]
+                
+        
+        
+    
+    def enviornmentFeedback(self, local_state, global_state, action, next_state, reward):
+        self.__updateSlidingWindows(local_state, action, reward)
+    
+        if (self.__isConflictDetected(local_state, action) and self.__isRewardLessThanAverage(local_state, action, reward)):
+            self.__markStateAsCoordinationNeeded(global_state)
+            
+        self.__updateQValues(local_state, global_state, action, next_state, reward)
+            
+    
+    def __updateSlidingWindows(self, local_state, action, reward):
+        if len(self.initial_rewards[local_state][action]) < self.sliding_window_size:
+            self.initial_rewards[local_state][action].append(reward)
+        else:
+            self.latest_rewards[local_state][action].append(reward)        
+    
+    def __isConflictDetected(self, local_state, action):
+        test_result = stats.ttest_ind(self.initial_rewards[local_state][action], self.latest_rewards[local_state][action])
+        return test_result.pvalue < .05
+    
+    def __isRewardLessThanAverage(self, local_state, action, reward):
+        return stats.ttest_1samp(self.latest_rewards[local_state][action], popmean=reward)
```

Comment:
@RadwaElaraby I think that there might be a bug here. I think that `ttest_1samp` returns a python object with `statistic` and `pvalue` rather than returning a boolean. I think that `environmentFeedback` is expecting it to return a boolean, though.

Reply:
I don't think it'll be a big issue, though - we'll just be a bit more likely to mark states as requiring coordination